examples/hello-world-xul/main.js

```diff
@@ -0,0 +1,34 @@
+/* Any copyright is dedicated to the Public Domain.
+ * http://creativecommons.org/publicdomain/zero/1.0/ */
+
+'use strict';
+
+const { classes: Cc, interfaces: Ci, results: Cr, utils: Cu } = Components;
+const { console } = Cu.import('resource://gre/modules/Console.jsm', {});
+const { Runtime } = Cu.import('resource://qbrt/modules/Runtime.jsm', {});
+const { Services } = Cu.import('resource://gre/modules/Services.jsm', {});
+
+const WINDOW_URL = 'chrome://app/content/index.html';
+const SHELL_URL = 'chrome://app/content/shell.xul';
+
+const WINDOW_FEATURES = [
+  'chrome',
+  'dialog=no',
+  'all',
+  'width=640',
+  'height=480',
+].join(',');
+
+// On startup, activate ourselves, since starting up from Node doesn't do this.
+// TODO: do this by default for all apps started via Node.
+if (Services.appinfo.OS === 'Darwin') {
+  Cc['@mozilla.org/widget/macdocksupport;1'].getService(Ci.nsIMacDockSupport).activateApplication(true);
+}
+
+const argument = Cc['@mozilla.org/supports-string;1'].createInstance(Ci.nsISupportsString);
+argument.data = WINDOW_URL;
+
+// TODO: report error if URL isn't found.
+const window = Services.ww.openWindow(null, SHELL_URL, '_blank', WINDOW_FEATURES, argument);
+Runtime.openDevTools(window);
+console.log('main.js loaded!');
```

Comment:
This is currently demonstrating how to build a XUL "shell" for an HTML app (i.e. the shell app). I would simplify it to focus on the use case of building a XUL app itself, i.e. to show the XUL equivalent of the hello-world HTML app.

Thus I would make main.js simply open a XUL window called index.xul, while index.xul would simply display "Hello, World!" in that window. Of course there are other features of XUL that this app could demonstrate, but simply displaying "Hello, World!" would be a good way to begin.

(It's also helpful for the two hello-world apps to have equivalent functionality, to the extent possible, so a user can look at them side-by-side to see the difference between developing in HTML vs. XUL. So if we add features to the XUL app in the future, then we should try to add the same features to the HTML app.)